File: test/modules/linear_boltzmann_solvers/transport_steady_cyl/transport_2d_cyl_1_monoenergetic.lua
-- 2D transport test in axialsymmetric cylindrical geometry with
-- vacuum boundary condition - monoenergetic.
-- SDM: PWLD
-- Test: Max-value=1.00000
num_procs = 4
--Structured mesh

-- Check num_procs
if check_num_procs == nil and number_of_processes ~= num_procs then
  log.Log(
    LOG_0ERROR,
    "Incorrect amount of processors. "
      .. "Expected "
      .. tostring(num_procs)
      .. ". Pass check_num_procs=false to override if possible."
  )
  os.exit(false)
end

-- Setup mesh
dim = 2
length = { 1.0, 2.0 }
ncells = { 50, 100 }
nodes = {}
for d = 1, dim do
  delta = length[d] / ncells[d]
  nodes[d] = {}
  for i = 0, ncells[d] do
    nodes[d][i + 1] = i * delta
  end
end

meshgen1 = mesh.OrthogonalMeshGenerator.Create({ node_sets = { nodes[1], nodes[2] } })
grid = meshgen1:Execute()

-- Set Material IDs
vol0 = logvol.RPPLogicalVolume.Create({
  xmin = 0.0,
  xmax = length[1],
  ymin = 0.0,
  ymax = length[2],
  infz = true,
})
grid:SetMaterialIDFromLogicalVolume(vol0, 0, true)

-- Add materials
ngrp = 1
sigmat = 25.0
ratioc = 0.1
source = sigmat * (1 - ratioc)

material0 = mat.AddMaterial("Material_0")
xs_1g = xs.CreateSimpleOneGroup(sigmat, ratioc)
material0:SetTransportXSections(xs_1g)
mg_src = lbs.VolumetricSource.Create({ block_ids = { 0 }, group_strength = { source } })

-- Setup Physics
pquad0 = aquad.GLCProductQuadrature2DRZ.Create({ Npolar = 4, Nazimuthal = 8 })

lbs_block = {
  mesh = grid,
  coord_system = 2,
  num_groups = ngrp,
  groupsets = {
    {
      groups_from_to = { 0, ngrp - 1 },
      angular_quadrature = pquad0,
      angle_aggregation_type = "azimuthal",
      inner_linear_method = "petsc_gmres",
      l_max_its = 100,
      l_abs_tol = 1.0e-12,
    },
  },
}

lbs_options = {
  boundary_conditions = { { name = "xmin", type = "reflecting" } },
  scattering_order = 0,
  volumetric_sources = { mg_src },
}
phys1 = lbs.DiscreteOrdinatesCurvilinearSolver.Create(lbs_block)
phys1:SetOptions(lbs_options)

-- Initialize and Execute Solver
ss_solver = lbs.SteadyStateSolver.Create({ lbs_solver = phys1 })

ss_solver:Initialize()
ss_solver:Execute()

--phys0 = LBSCurvilinearCreateSolver(LBSCurvilinear.CYLINDRICAL)
--
----  angular quadrature
--pquad = aquad.CreateCylindricalProductQuadrature(GAUSS_LEGENDRE_CHEBYSHEV, 4, 8)
--
----  groups
--groups = {}
--for g = 1, ngrp do
--  groups[g] = LBSCreateGroup(phys0)
--end
--
----  groupsets
--gs0 = LBSCreateGroupset(phys0)
--LBSGroupsetAddGroups(phys0, gs0, 0, ngrp-1)
--LBSGroupsetSetQuadrature(phys0, gs0, pquad)
--LBSGroupsetSetAngleAggregationType(phys0, gs0, LBSGroupset.ANGLE_AGG_AZIMUTHAL)
--LBSGroupsetSetIterativeMethod(phys0, gs0, KRYLOV_GMRES_CYCLES)
--LBSGroupsetSetResidualTolerance(phys0, gs0, 1.0e-12)
--LBSGroupsetSetMaxIterations(phys0, gs0, 100)
--LBSGroupsetSetGMRESRestartIntvl(phys0, gs0, 30)
--
----  spatial discretisation
--LBSSetProperty(phys0, DISCRETIZATION_METHOD, PWLD)
--
----  scattering order
--LBSSetProperty(phys0, SCATTERING_ORDER, 0)
--
----------------------------------------------------------------------------------
----  boundary conditions
----------------------------------------------------------------------------------
--dirichlet_value = {}
--for g = 1, ngrp do
--  dirichlet_value[g] = 0
--end
--LBSSetProperty(phys0, BOUNDARY_CONDITION,
--                  XMIN, LBSBoundaryTypes.REFLECTING)
--LBSSetProperty(phys0, BOUNDARY_CONDITION,
--                  XMAX, LBSBoundaryTypes.INCIDENT_ISOTROPIC, dirichlet_value)
--LBSSetProperty(phys0, BOUNDARY_CONDITION,
--                  YMIN, LBSBoundaryTypes.INCIDENT_ISOTROPIC, dirichlet_value)
--LBSSetProperty(phys0, BOUNDARY_CONDITION,
--                  YMAX, LBSBoundaryTypes.INCIDENT_ISOTROPIC, dirichlet_value)
--
----------------------------------------------------------------------------------
----  solvers
----------------------------------------------------------------------------------
--solver.Initialize(phys0)
--solver.Execute(phys0)

-- Exports
fflist = lbs.GetScalarFieldFunctionList(phys1)
if master_export == nil then
  fieldfunc.ExportToVTKMulti(fflist, "ZRZPhi")
end

-- Volume integrations
ffi1 = fieldfunc.FieldFunctionInterpolationVolume.Create()
curffi = ffi1
curffi:SetOperationType(OP_MAX)
curffi:SetLogicalVolume(vol0)
curffi:AddFieldFunction(fflist[1])

curffi:Initialize()
curffi:Execute()
maxval = curffi:GetValue()

log.Log(LOG_0, string.format("Max-value=%.5f", maxval))

File: test/python/modules/linear_boltzmann_solvers/transport_steady_cyl/transport_2d_cyl_1_monoenergetic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
2D monoenergetic transport test in axialsymmetric cylindrical geometry with
vacuum boundary conditions

Test: Max-value=1.00000
"""

import os
import sys
import math

if "opensn_console" not in globals():
    from mpi4py import MPI

    size = MPI.COMM_WORLD.size
    rank = MPI.COMM_WORLD.rank
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import OrthogonalMeshGenerator
    from pyopensn.logvol import RPPLogicalVolume
    from pyopensn.xs import MultiGroupXS
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLCProductQuadrature2DRZ
    from pyopensn.solver import DiscreteOrdinatesCurvilinearProblem, SteadyStateSolver
    from pyopensn.fieldfunc import FieldFunctionInterpolationVolume

if __name__ == "__main__":

    num_procs = 4
    if size != num_procs:
        sys.exit(f"Incorrect number of processors. Expected {num_procs} but got {size}.")

    # Setup mesh
    dim = 2
    length = [1.0, 2.0]
    ncells = [50, 100]
    nodes = []
    for d in range(dim):
        delta = length[d] / ncells[d]
        node_set = [i * delta for i in range(ncells[d] + 1)]
        nodes.append(node_set)
    meshgen = OrthogonalMeshGenerator(node_sets=[nodes[0], nodes[1]], coord_sys="cylindrical")
    grid = meshgen.Execute()

    # Set block IDs
    vol0 = RPPLogicalVolume(
        xmin=0.0,
        xmax=length[0],
        ymin=0.0,
        ymax=length[1],
        infz=True,
    )
    grid.SetBlockIDFromLogicalVolume(vol0, 0, True)

    # Define materials and sources (one-group problem)
    ngrp = 1
    sigmat = 25.0
    ratioc = 0.1
    xs_1g = MultiGroupXS()
    xs_1g.CreateSimpleOneGroup(sigmat, ratioc)
    source = []
    source.append(sigmat * (1 - ratioc))
    mg_src = VolumetricSource(block_ids=[0], group_strength=source)

    # Angular quadrature
    pquad = GLCProductQuadrature2DRZ(n_polar=4, n_azimuthal=8, scattering_order=0)

    # Create the curvilinear solver (for cylindrical geometry)
    phys = DiscreteOrdinatesCurvilinearProblem(
        mesh=grid,
        num_groups=ngrp,
        groupsets=[
            {
                "groups_from_to": (0, ngrp - 1),
                "angular_quadrature": pquad,
                "angle_aggregation_type": "azimuthal",
                "inner_linear_method": "petsc_gmres",
                "l_max_its": 100,
                "l_abs_tol": 1.0e-12,
            },
        ],
        xs_map=[
            {"block_ids": [0], "xs": xs_1g},
        ],
        scattering_order=0,
        options={
            "boundary_conditions": [{"name": "xmin", "type": "reflecting"}],
            "volumetric_sources": [mg_src],
        }
    )
    ss_solver = SteadyStateSolver(lbs_problem=phys)
    ss_solver.Initialize()
    ss_solver.Execute()

    # Field functions
    fflist = phys.GetScalarFieldFunctionList(only_scalar_flux=False)

    # Perform a volume integration over vol0 (compute the maximum value)
    ffi1 = FieldFunctionInterpolationVolume()
    curffi = ffi1
    curffi.SetOperationType("max")
    curffi.SetLogicalVolume(vol0)
    curffi.AddFieldFunction(fflist[0][0])
    curffi.Initialize()
    curffi.Execute()
    maxval = curffi.GetValue()
    if rank == 0:
        print(f"Max-value={maxval:.5f}")
